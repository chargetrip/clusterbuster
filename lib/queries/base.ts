import { GetBaseQuery, IBaseQueryInput } from '../types/index';

/**
 * @description The default base query builder
 */
export const defaultGetBaseQuery: GetBaseQuery = ({
  x,
  y,
  z,
  table,
  geometry,
  maxZoomLevel,
  attributes,
  query,
}: IBaseQueryInput) => `
SELECT
st_flipcoordinates(${geometry}) AS center,
  1 AS size,
  0 AS clusterNo,
  ${maxZoomLevel + 1} AS expansionZoom${attributes}
FROM ${table}
WHERE 
	ST_Intersects(TileBBox(${z}, ${x}, ${y}, 3857), ST_Transform(${geometry}, 3857))
	${query.length > 0 ? `AND ${query.join(' AND ')}` : ''}
`;
