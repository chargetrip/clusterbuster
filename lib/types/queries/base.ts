/**
 * @description Input interface for the base query builder
 */
export interface IBaseQueryInput {
  x: number;
  y: number;
  z: number;
  table: string;
  geometry: string;
  maxZoomLevel: number;
  attributes: string;
  query: string[];
}

/**
 * @description The base query builder callback definition
 */
export type GetBaseQuery = (input: IBaseQueryInput) => string;
