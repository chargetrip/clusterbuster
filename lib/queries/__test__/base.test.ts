import { defaultGetBaseQuery } from '../base';

describe('defaultGetBaseQuery', () => {
  it('should work', () => {
    expect(
      defaultGetBaseQuery({
        x: 0,
        y: 0,
        z: 0,
        table: 'table',
        geometry: 'geometry',
        maxZoomLevel: 12,
        attributes: ', id, name',
        query: [],
      })
    ).toMatchSnapshot();
    expect(
      defaultGetBaseQuery({
        x: 0,
        y: 0,
        z: 0,
        table: 'table',
        geometry: 'geometry',
        maxZoomLevel: 12,
        attributes: ', id, name',
        query: ["name = 'name'"],
      })
    ).toMatchSnapshot();
    expect(
      defaultGetBaseQuery({
        x: 0,
        y: 0,
        z: 0,
        table: 'table',
        geometry: 'geometry',
        maxZoomLevel: 12,
        attributes: ', id, name',
        query: ["name = 'name'", "id = '1'"],
      })
    ).toMatchSnapshot();
  });
});
