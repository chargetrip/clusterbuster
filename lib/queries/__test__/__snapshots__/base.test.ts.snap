// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultGetBaseQuery should work 1`] = `
"
SELECT
  geometry AS center,
  1 AS size,
  0 AS clusterNo,
  13 AS expansionZoom, id, name
FROM table
WHERE 
	ST_Intersects(TileBBox(0, 0, 0, 3857), ST_Transform(geometry, 3857))
	
"
`;

exports[`defaultGetBaseQuery should work 2`] = `
"
SELECT
  geometry AS center,
  1 AS size,
  0 AS clusterNo,
  13 AS expansionZoom, id, name
FROM table
WHERE 
	ST_Intersects(TileBBox(0, 0, 0, 3857), ST_Transform(geometry, 3857))
	AND name = 'name'
"
`;

exports[`defaultGetBaseQuery should work 3`] = `
"
SELECT
  geometry AS center,
  1 AS size,
  0 AS clusterNo,
  13 AS expansionZoom, id, name
FROM table
WHERE 
	ST_Intersects(TileBBox(0, 0, 0, 3857), ST_Transform(geometry, 3857))
	AND name = 'name' AND id = '1'
"
`;
